### ISSUE TITLE
"register" variable not always optimized

### ISSUE THREAD

**Issue by Senbei64** [cc65/cc65#209]
Simple source code to show the issue:

	char c, r;
	void main(void)
	{
		register char * p;
		c = p[0x10];
		r = 0x20;
		c = p[r];
	}

the statement "c = p[10]" gets compiled as:

	ldy     #$0A
	lda     (regbank+4),y
	sta     _c

but the statement "c = p[r]" gets compiled as:

	lda     regbank+4
	ldx     regbank+4+1
	ldy     _r
	sta     ptr1
	stx     ptr1+1
	lda     (ptr1),y
	sta     _c

where one might expect instead:

	ldy     _r
	lda     (regbank+4),y
	sta     _c

**Comment by MonteCarlos** [comment]
Yeah, thats why it is not recommended to turn on register variables 
generally...
Seems not easy to fix that as i also posted this issue several times and
there was no suggestion other than to refrain from using register 
variables .


Am 15.09.2015 um 19:21 schrieb Senbei64:

**Comment by mrdudz** [comment]
in the meantime, this became worse for some reason:
```
.segment        "CODE"

;
; register char * p;
;
        lda     regbank+4
        ldx     regbank+5
        jsr     pushax
;
; c = p[10];
;
        lda     regbank+4
        ldx     regbank+4+1
        ldy     #$0A
        sta     ptr1
        stx     ptr1+1
        lda     (ptr1),y
        sta     _c
;
; r = 20;
;
        lda     #$14
        sta     _r
;
; c = p[r];
;
        lda     regbank+4
        ldx     regbank+4+1
        ldy     _r
        sta     ptr1
        stx     ptr1+1
        lda     (ptr1),y
        sta     _c
;
; }
;
        ldy     #$00
        lda     (sp),y
        sta     regbank+4
        iny
        lda     (sp),y
        sta     regbank+5
        jmp     incsp2
```

**Comment by mrdudz** [comment]
got slightly better again (saving/restoring register bank gets optimized out)
```
.proc	_main: near

.segment	"CODE"

;
; c = p[10];
;
	lda     regbank+4
	ldx     regbank+4+1
	ldy     #$0A
	sta     ptr1
	stx     ptr1+1
	lda     (ptr1),y
	sta     _c
;
; r = 20;
;
	lda     #$14
	sta     _r
;
; c = p[r];
;
	lda     regbank+4
	ldx     regbank+4+1
	ldy     _r
	sta     ptr1
	stx     ptr1+1
	lda     (ptr1),y
	sta     _c
;
; }
;
	rts
```

**Comment by kugelfuhr** [comment]
> got slightly better again (saving/restoring register bank gets optimized out)

That's only for the `main()` function, not for the general case.

The regression was introduced when merging `ArrayRef()` into `parseadd()` quite some time ago. This caused array references to be translated to

```asm
;
; c = p[10];
;
        .dbg    line, "t.c", 5
        lda     regbank+4
        ldx     regbank+4+1
        ldy     #$0A
        jsr     incaxy
        ldy     #$00
        jsr     ldauidx
        sta     _c
```

instead of

```asm
;
; c = p[10];
;
        .dbg    line, "t.c", 5
        lda     regbank+4
        ldx     regbank+4+1
        ldy     #$0A
        jsr     ldauidx
        sta     _c
```

which made a whole code chain in the optimizer handling these sequences more or less worthless.

Unfortunately I see no quick solution to get the old code generation back. It is buried somewhere inside the 400 lines of `parseadd()`. The optimizations are still triggered if the offset is zero, but not for other offsets. This means it is quite often not a good idea to use register variables. Their most effective use was for pointers to arrays but is no longer the case.

**Comment by kugelfuhr** [comment]
> in the meantime, this became worse for some reason:

Just for the record: there's already #2055 which looks like it has the same cause.

There are so many issues open (a lot of them useless) that it is difficult to keep the overview.
